File: _2020/go/days/day2.go
package days

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/binnev/advent/utils"
)

func parseLine(line string) (string, string, string, string) {
	rx := regexp.MustCompile(`(?P<a>\d+)-(?P<b>\d+) (?P<c>\w+): (?P<d>\w+)`)
	match := rx.FindStringSubmatch(line)
	a := match[1]
	b := match[2]
	char := match[3]
	pwd := match[4]
	return a, b, char, pwd
}

func Day2Part1() string {
	data := utils.LoadPuzzleInput("day2")
	lines := strings.Split(data, "\n")
	num_valid := 0
	for _, line := range lines {
		a, b, char, pwd := parseLine(line)
		min, _ := strconv.Atoi(a)
		max, _ := strconv.Atoi(b)
		count := strings.Count(pwd, char)
		if min <= count && count <= max {
			num_valid += 1
		}
	}
	return fmt.Sprintf("%v", num_valid)
}

func Day2Part2() string {
	data := utils.LoadPuzzleInput("day2")
	lines := strings.Split(data, "\n")
	num_valid := 0
	for _, line := range lines {
		a, b, char, pwd := parseLine(line)
		index1, _ := strconv.Atoi(a)
		index2, _ := strconv.Atoi(b)
		if (string(pwd[index1-1]) == char) != (string(pwd[index2-1]) == char) {
			num_valid += 1
		}
	}
	return fmt.Sprintf("%v", num_valid)
}

func Day2() {
	utils.Profile(Day2Part1)
	utils.Profile(Day2Part2)
}

File: python/puzzles/_2020/day2.py
from utils import load_puzzle_input, profile

puzzle_input = load_puzzle_input("2020/day2")


def parse_line(line: str) -> tuple[int, int, str, str]:
    policy, password = line.split(": ")
    numbers, letter = policy.split()
    a, b = map(int, numbers.split("-"))
    return a, b, letter, password


@profile
def part1():
    num_valid = 0
    for line in puzzle_input.split("\n"):
        _min, _max, letter, password = parse_line(line)
        if _min <= password.count(letter) <= _max:
            num_valid += 1
    return num_valid


@profile
def part2():
    num_valid = 0
    for line in puzzle_input.split("\n"):
        index1, index2, letter, password = parse_line(line)
        if (password[index1 - 1] == letter) ^ (password[index2 - 1] == letter):
            num_valid += 1
    return num_valid


if __name__ == "__main__":
    part1()
    part2()
